### DISCUSSION TITLE
Weird constraint name behavior

### DISCUSSION THREAD

**Discussion by skyf0l** [planetscale/discussion#613] [Bug Reports]
As you can see, the schemas only differ in the constraint names, but it's impossible to sync them on my main branch. Is this due to planetscale's behavior? How can I rename the constraints in my main branch? I've tried everything with deploy requests, but it's impossible to succeed.

The problem is that this results in a non-empty schema diff and makes my continuous deployment process fail...

![image](https://github.com/planetscale/discussion/assets/59019720/07a6b3f8-5904-4a8a-9441-30d87cb77c09)

![image](https://github.com/planetscale/discussion/assets/59019720/fdbdea20-6a9a-4a2e-9830-0cf0f8fe6e69)

**Comment by notfelineit** [comment]
Hi @skyf0l , would you mind sharing your database branch name, database name, and organization? That would greatly help us debug this issue for you 😄

**Reply by skyf0l** [reply]
Orga: `nema`
Database: `nema`
Main branch: `stable`
Dev branch: `pr-[0-9]+`

**Reply by notfelineit** [reply]
Thanks @skyf0l , we'll get back to you shortly!

**Comment by notfelineit** [comment]
Just an update - we were able to recreate this issue ourselves, and will be thinking through a few options on Monday!

**Comment by notfelineit** [comment]
Hi @skyf0l , we added documentation to capture the "renaming of constraint names" here: https://planetscale.com/docs/concepts/foreign-key-constraints#foreign-key-constraint-names-change-on-every-deployment

**TLDR:** because of a mysql limitation that constraint names must be unique to the database schema, combined with the way our online DDL works, we must change constraint names if the table in which the constraint lies is modified by a DR.

However, there is a bug, in that you _should not_ be seeing a diff in your dev branch's schema:
![image](https://github.com/planetscale/discussion/assets/31225471/6a1852d0-d0c0-40db-985f-8706502b08b7)

We'll be fixing this, and the _correct state_ is indeed the "There are no schema changes in this branch":
![image](https://github.com/planetscale/discussion/assets/31225471/7db13d35-ba10-4050-b4dd-2327a09a55ad)

**Comment by skyf0l** [comment]
Hello @notfelineit, from the front view, it's fixed, but the `pscale` cli always detects a change.

![image](https://github.com/planetscale/discussion/assets/59019720/dd610af5-8329-4483-8d22-182373bf1032)

This happen with only new constraint names

```console
$ pscale branch diff nema --org nema -f json test
[
  {
    "name": "Account",
    "raw": " CREATE TABLE `Account` (\n \t`userId` varchar(36) NOT NULL,\n \t`provider` varchar(255) NOT NULL,\n \t`providerAccountId` varchar(255) NOT NULL,\n \t`email` varchar(255),\n \t`username` varchar(255),\n \t`mfaEnabled` tinyint(1),\n \t`emailVerified` tinyint(1),\n \tPRIMARY KEY (`provider`, `providerAccountId`),\n \tKEY `Account_userId_User_id_fk` (`userId`),\n-\tCONSTRAINT `Account_userId_User_id_fk_0ragpof0t4yv849kzumrjomv6` FOREIGN KEY (`userId`) REFERENCES `User` (`id`) ON DELETE CASCADE\n+\tCONSTRAINT `Account_userId_User_id_fk` FOREIGN KEY (`userId`) REFERENCES `User` (`id`) ON DELETE CASCADE\n ) ENGINE InnoDB,\n   CHARSET utf8mb4,\n   COLLATE utf8mb4_0900_ai_ci;\n",
    "html": "\u003cdiv class='diff'\u003e\u003cul\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\u003cspan class=\"k\"\u003eCREATE\u003c/span\u003e \u003cspan class=\"k\"\u003eTABLE\u003c/span\u003e \u003cspan class=\"nv\"\u003e`Account`\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`userId`\u003c/span\u003e \u003cspan class=\"nb\"\u003evarchar\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e36\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eNOT\u003c/span\u003e \u003cspan class=\"k\"\u003eNULL\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`provider`\u003c/span\u003e \u003cspan class=\"nb\"\u003evarchar\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e255\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eNOT\u003c/span\u003e \u003cspan class=\"k\"\u003eNULL\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`providerAccountId`\u003c/span\u003e \u003cspan class=\"nb\"\u003evarchar\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e255\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eNOT\u003c/span\u003e \u003cspan class=\"k\"\u003eNULL\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`email`\u003c/span\u003e \u003cspan class=\"nb\"\u003evarchar\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e255\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`username`\u003c/span\u003e \u003cspan class=\"nb\"\u003evarchar\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e255\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`mfaEnabled`\u003c/span\u003e \u003cspan class=\"nb\"\u003etinyint\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e1\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"nv\"\u003e`emailVerified`\u003c/span\u003e \u003cspan class=\"nb\"\u003etinyint\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e1\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"k\"\u003ePRIMARY\u003c/span\u003e \u003cspan class=\"k\"\u003eKEY\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`provider`\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"nv\"\u003e`providerAccountId`\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\t\u003cspan class=\"k\"\u003eKEY\u003c/span\u003e \u003cspan class=\"nv\"\u003e`Account_userId_User_id_fk`\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`userId`\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\u003c/li\u003e\n\u003cli class='del'\u003e- \t\u003cspan class=\"k\"\u003eCONSTRAINT\u003c/span\u003e \u003cspan class=\"nv\"\u003e`Account_userId_User_id_fk_0ragpof0t4yv849kzumrjomv6`\u003c/span\u003e \u003cspan class=\"k\"\u003eFOREIGN\u003c/span\u003e \u003cspan class=\"k\"\u003eKEY\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`userId`\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eREFERENCES\u003c/span\u003e \u003cspan class=\"nv\"\u003e`User`\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`id`\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eON\u003c/span\u003e \u003cspan class=\"k\"\u003eDELETE\u003c/span\u003e \u003cspan class=\"k\"\u003eCASCADE\u003c/span\u003e\u003c/li\u003e\n\u003cli class='ins'\u003e+ \t\u003cspan class=\"k\"\u003eCONSTRAINT\u003c/span\u003e \u003cspan class=\"nv\"\u003e`Account_userId_User_id_fk`\u003c/span\u003e \u003cspan class=\"k\"\u003eFOREIGN\u003c/span\u003e \u003cspan class=\"k\"\u003eKEY\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`userId`\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eREFERENCES\u003c/span\u003e \u003cspan class=\"nv\"\u003e`User`\u003c/span\u003e \u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"nv\"\u003e`id`\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003eON\u003c/span\u003e \u003cspan class=\"k\"\u003eDELETE\u003c/span\u003e \u003cspan class=\"k\"\u003eCASCADE\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"n\"\u003eENGINE\u003c/span\u003e \u003cspan class=\"n\"\u003eInnoDB\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;  \u003cspan class=\"n\"\u003eCHARSET\u003c/span\u003e \u003cspan class=\"n\"\u003eutf8mb4\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003c/li\u003e\n\u003cli class='unchanged'\u003e\u0026nbsp;  \u003cspan class=\"k\"\u003eCOLLATE\u003c/span\u003e \u003cspan class=\"n\"\u003eutf8mb4_0900_ai_ci\u003c/span\u003e\u003cspan class=\"p\"\u003e;\u003c/span\u003e\u003c/li\u003e\n\u003c/ul\u003e\u003c/div\u003e"
  }
]
```

And this with same constraint names:
```console
$ pscale branch diff nema --org nema -f json test
[]
```

**Reply by nick-rackauckas** [reply]
@skyf0l  Still seeing this issue on the front view right now after just creating a deploy request

Organization: usable-machines
Database: um-core
DR #241

**Reply by skyf0l** [reply]
Hello @notfelineit, are there any updates on this issue?

**Reply by notfelineit** [reply]
Hey, I'll check the status! We still need to update our diff generator. Functionally, things should be the same.

**Comment by nick-rackauckas** [comment]
Still seeing this issue consistently 
Organization: usable-machines
Database: um-core
DR examples: 241, 247, and pretty much any other ones around this time

**Reply by notfelineit** [reply]
Hey Nick - I want to clarify that the constraints being renamed is a known behavior that _will_ happen because of our online DDL mechanism. 

The bug is specifically that the diff shows. Are you still seeing the diff?

**Reply by nick-rackauckas** [reply]
> Hey Nick - I want to clarify that the constraints being renamed is a known behavior that _will_ happen because of our online DDL mechanism. 
> 
> 
> 
> The bug is specifically that the diff shows. Are you still seeing the diff? 

Correct, in nearly all our Deploy Requests

**Reply by skyf0l** [reply]
Same for me, because we automatically delete dev branches that have no changes with the prod branch, and now nothing happens because the schema changes due to constraint names.

**Comment by skyf0l** [comment]
Hello, any updates? now that no one is taking advantage of the free tier and has to pay, I think it would be interesting to push this kind of issue a little faster if possible.

(front view is broken again)

![image](https://github.com/planetscale/discussion/assets/59019720/4abc4d51-d883-4d1c-83d5-c917ba61d5a9)

**Comment by skyf0l** [comment]
Hello @notfelineit, are there any updates on this issue? I'm still seeing this bug in all our deploy requests...
